### ISSUE TITLE
False Positive on E1121 for Usage of Self

### ISSUE THREAD

**Issue by nga-27** [pylint-dev/pylint#9239]
### Bug description

```python
# file a.py
from typing import Callable, Union

class A:
   def do_stuff(self, data: bytes) -> Union[None, str]:
       for map_name, x_function in self.x_functions.items():
           return x_function(self, data)

   def funct_1(self, data: bytes) -> None:
       # something
       return None

   def funct_2(self, data: bytes) -> str:
       return str(bytes)

   x_functions: dict[str, Callable[[Self, bytes], Union[None, str]] = {
       "funct_1": funct_1,
       "funct_2": funct_2
   }
```

### Command used

```shell
pylint a.py
```

### Pylint output

```python
a.py:7:17: E1121: Too many positional arguments for method call (too-many-function-args)
```

### Expected behavior

My assumption would be that this is a false positive, as "self" has to be included in the generic function call on line 7 (above). Without "self", the script fails at runtime due to not recognizing the function call "funct_1" (and, therefore, "funct_2").

It also looks like explicitly adding "Self" type from `typing` to the x_functions dictionary doesn't help the linter.

I understand this is an odd way to do such a function, but in the real function I'm implementing, I'm comparing packet preambles that are variable length, thus the `for name in names` comprehension.

### Pylint version

```shell
pylint 3.0.2
```

### OS / Environment

Windows 10

**Comment by msalim-eightfold** [comment]
facing the same issue
